src/cmap/auth/mongo_credentials.ts

```diff
@@ -108,4 +108,54 @@ export class MongoCredentials {
 
     return this;
   }
+
+  validate(): void {
+    if (
+      (this.mechanism === AuthMechanism.MONGODB_GSSAPI ||
+        this.mechanism === AuthMechanism.MONGODB_CR ||
+        this.mechanism === AuthMechanism.MONGODB_PLAIN ||
+        this.mechanism === AuthMechanism.MONGODB_SCRAM_SHA1 ||
+        this.mechanism === AuthMechanism.MONGODB_SCRAM_SHA256) &&
+      !this.username
+    ) {
+      throw new TypeError(`Username required for mechanism '${this.mechanism}'`);
+    }
+
+    if (
+      this.mechanism === AuthMechanism.MONGODB_GSSAPI ||
+      this.mechanism === AuthMechanism.MONGODB_AWS ||
+      this.mechanism === AuthMechanism.MONGODB_X509
+    ) {
+      if (this.source != null && this.source !== '$external') {
+        throw new TypeError(
+          `Invalid source '${this.source}' for mechanism '${this.mechanism}' specified.`
+        );
+      }
+    }
+
+    if (this.mechanism === AuthMechanism.MONGODB_PLAIN && this.source == null) {
+      throw new TypeError('PLAIN Authentication Mechanism needs an auth source');
+    }
+
+    if (this.mechanism === AuthMechanism.MONGODB_X509 && this.password != null) {
+      if (this.password === '') {
+        Reflect.set(this, 'password', undefined);
+        return;
+      }
+      throw new TypeError(`Password not allowed for mechanism MONGODB-X509`);
+    }
+  }
+
+  static merge(
+    creds: MongoCredentials,
+    options: Partial<MongoCredentialsOptions>
+  ): MongoCredentials {
+    return new MongoCredentials({
+      username: options.username ?? creds.username,
+      password: options.password ?? creds.password,
+      mechanism: options.mechanism ?? creds.mechanism,
+      mechanismProperties: options.mechanismProperties ?? creds.mechanismProperties,
+      source: options.source ?? creds.source ?? options.db
+    });
```

Comment:
@mbroadst 
Sorry about my git fu, I can't reply to your comment about the credentials issue. I've gone ahead with the merge solution, lmk what you think. For what its worth I think we were safe with the spread solution because typescript would not permit us to have a property missing, it just so happens that MongoCredentials and MongoCredentialOptions share all their properties and mandate that they are defined, but ofc better to be safe and box up more complex functionality, so `.merge` seems like a good solution .